File: matlab/stereoBP_Synch2.m
% Stereo Matching using Belief Propagation (with Synchronous message update schedule) - a different aproach
% Computes a disparity map from a rectified stereo pair using Belief Propagation

% Set parameters
dispLevels = 16; %disparity range: 0 to dispLevels-1
iterations = 60;
lambda = 5; %weight of smoothness cost

% Define data cost computation
dataCostComputation = @(left,right) abs(left-right); %absolute differences
%dataCostComputation = @(left,right) (left-right).^2; %square differences

% Predefined smoothness cost computation: lambda*min(abs(differences),2)

% Start timer
timerVal = tic();

% Load left and right images in grayscale
leftImg = rgb2gray(imread('left.png'));
rightImg = rgb2gray(imread('right.png'));

% Apply a Gaussian filter
leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

% Get the size
[rows,cols] = size(leftImg);

% Convert to int32
leftImg = int32(leftImg);
rightImg = int32(rightImg);

% Compute pixel-based matching cost (data cost)
dataCost = zeros(rows,cols,dispLevels,'int32');
for d = 0:dispLevels-1
    %rightImgShifted = shiftArray(rightImg,[0,d]);
    rightImgShifted = circshift(rightImg,d,2); %less accurate, better performances
    dataCost(:,:,d+1) = dataCostComputation(leftImg,rightImgShifted);
end

% Initialize messages
msgFromUp = zeros(rows,cols,dispLevels,'int32');
msgFromDown = zeros(rows,cols,dispLevels,'int32');
msgFromRight = zeros(rows,cols,dispLevels,'int32');
msgFromLeft = zeros(rows,cols,dispLevels,'int32');

msgToUp1 = intmax*ones(rows,cols,dispLevels+2,'int32');
msgToDown1 = intmax*ones(rows,cols,dispLevels+2,'int32');
msgToRight1 = intmax*ones(rows,cols,dispLevels+2,'int32');
msgToLeft1 = intmax*ones(rows,cols,dispLevels+2,'int32');

msgToUp2 = zeros(rows,cols,dispLevels,'int32');
msgToDown2 = zeros(rows,cols,dispLevels,'int32');
msgToRight2 = zeros(rows,cols,dispLevels,'int32');
msgToLeft2 = zeros(rows,cols,dispLevels,'int32');

costs = zeros(rows,cols,3,'int32');
energy = zeros(iterations,1);
figure

% Start iterations
for it = 1:iterations

    % Compute messages - Step 1
    msgToUp1(:,:,2:end-1) = dataCost + msgFromDown + msgFromRight + msgFromLeft;
    msgToDown1(:,:,2:end-1) = dataCost + msgFromUp + msgFromRight + msgFromLeft;
    msgToRight1(:,:,2:end-1) = dataCost + msgFromUp + msgFromDown + msgFromLeft;
    msgToLeft1(:,:,2:end-1) = dataCost + msgFromUp + msgFromDown + msgFromRight;

    % Find minimum costs
    minMsgToUp = min(msgToUp1,[],3);
    minMsgToDown = min(msgToDown1,[],3);
    minMsgToRight = min(msgToRight1,[],3);
    minMsgToLeft = min(msgToLeft1,[],3);

    % Compute messages - Step 2
    for i = 1:dispLevels
        % Messages to up
        costs(:,:,1) = msgToUp1(:,:,i+1);
        costs(:,:,2) = min(msgToUp1(:,:,i),msgToUp1(:,:,i+2))+lambda;
        costs(:,:,3) = minMsgToUp+2*lambda;
        msgToUp2(:,:,i) = min(costs,[],3)-minMsgToUp;
        
        % Messages to down
        costs(:,:,1) = msgToDown1(:,:,i+1);
        costs(:,:,2) = min(msgToDown1(:,:,i),msgToDown1(:,:,i+2))+lambda;
        costs(:,:,3) = minMsgToDown+2*lambda;
        msgToDown2(:,:,i) = min(costs,[],3)-minMsgToDown;
        
        % Messages to right
        costs(:,:,1) = msgToRight1(:,:,i+1);
        costs(:,:,2) = min(msgToRight1(:,:,i),msgToRight1(:,:,i+2))+lambda;
        costs(:,:,3) = minMsgToRight+2*lambda;
        msgToRight2(:,:,i) = min(costs,[],3)-minMsgToRight;
        
        % Messages to left
        costs(:,:,1) = msgToLeft1(:,:,i+1);
        costs(:,:,2) = min(msgToLeft1(:,:,i),msgToLeft1(:,:,i+2))+lambda;
        costs(:,:,3) = minMsgToLeft+2*lambda;
        msgToLeft2(:,:,i) = min(costs,[],3)-minMsgToLeft;
    end

    % Send messages
    msgFromDown = shiftArray(msgToUp2,[-1,0,0]); %shift up
    msgFromUp = shiftArray(msgToDown2,[1,0,0]); %shift down
    msgFromLeft = shiftArray(msgToRight2,[0,1,0]); %shift right
    msgFromRight = shiftArray(msgToLeft2,[0,-1,0]); %shift left

    % Compute belief
    %belief = dataCost + msgFromUp + msgFromDown + msgFromRight + msgFromLeft; %standard belief computation
    belief = msgFromUp + msgFromDown + msgFromRight + msgFromLeft; %without dataCost (larger energy but better results)
    
    % Compute the disparity map
    [~,ind] = min(belief,[],3);
    dispMap = ind-1;
    
    % Compute energy
    [row,col] = ndgrid(1:size(ind,1),1:size(ind,2));
    linInd = sub2ind(size(dataCost),row,col,ind);
    dataEnergy = sum(sum(dataCost(linInd)));
    smoothnessEnergyHorizontal = sum(sum(lambda*min(abs(diff(dispMap,1,2)),2)));
    smoothnessEnergyVertical = sum(sum(lambda*min(abs(diff(dispMap,1,1)),2)));
    energy(it) = dataEnergy+smoothnessEnergyHorizontal+smoothnessEnergyVertical;

    % Normalize the disparity map for display
    scaleFactor = 256/dispLevels;
    dispImg = uint8(dispMap*scaleFactor);

    % Show disparity map
    imshow(dispImg)
    
    % Show energy and iteration
    fprintf('iteration: %d/%d, energy: %d\n',it,iterations,energy(it))
end

% Show convergence graph
figure
plot(1:iterations,energy,'bo-')
xlabel('Iterations')
ylabel('Energy')

% Save disparity map
imwrite(dispImg,'disparityBP_Synch2.png')

% Stop timer and display running time
elapsedTime = toc(timerVal);
fprintf('Running time: %.2f seconds\n',elapsedTime)

File: python/stereoBP_Synch2.py
# Stereo Matching using Belief Propagation (with Synchronous message update schedule) - a different aproach
# Computes a disparity map from a rectified stereo pair using Belief Propagation

import time
import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt
from shiftArray import shiftArray

MAX_INT = 2147483647

# Set parameters
dispLevels = 16 #disparity range: 0 to dispLevels-1
iterations = 60
lambda_ = 5 #weight of smoothness cost

# Define data cost computation
dataCostComputation = lambda left,right: np.absolute(left-right) #absolute differences
#dataCostComputation = lambda left,right: (left-right)**2 #square differences

# Predefined smoothness cost computation: lambda_*np.minimum(np.absolute(differences),2)

# Start timer
timerVal = time.time()

# Load left and right images in grayscale
leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

# Apply a Gaussian filter
leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

# Get the size
(rows,cols) = leftImg.shape

# Convert to int32
leftImg = leftImg.astype(np.int32)
rightImg = rightImg.astype(np.int32)

# Compute pixel-based matching cost (data cost)
dataCost = np.zeros((rows,cols,dispLevels),dtype=np.int32)
for d in range(dispLevels):
    #rightImgShifted = shiftArray(rightImg,[0,d])
    rightImgShifted = np.roll(rightImg,d,1) #less accurate, better performances
    dataCost[:,:,d] = dataCostComputation(leftImg,rightImgShifted)

# Initialize messages
msgFromUp = np.zeros((rows,cols,dispLevels),dtype=np.int32)
msgFromDown = np.zeros((rows,cols,dispLevels),dtype=np.int32)
msgFromRight = np.zeros((rows,cols,dispLevels),dtype=np.int32)
msgFromLeft = np.zeros((rows,cols,dispLevels),dtype=np.int32)

msgToUp1 = MAX_INT*np.ones((rows,cols,dispLevels+2),dtype=np.int32)
msgToDown1 = MAX_INT*np.ones((rows,cols,dispLevels+2),dtype=np.int32)
msgToRight1 = MAX_INT*np.ones((rows,cols,dispLevels+2),dtype=np.int32)
msgToLeft1 = MAX_INT*np.ones((rows,cols,dispLevels+2),dtype=np.int32)

msgToUp2 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
msgToDown2 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
msgToRight2 = np.zeros((rows,cols,dispLevels),dtype=np.int32)
msgToLeft2 = np.zeros((rows,cols,dispLevels),dtype=np.int32)

costs = np.zeros((rows,cols,3),dtype=np.int32)
energy = np.zeros(iterations,dtype=np.int32)

# Start iterations
for it in range(iterations):

    # Compute messages - Step 1
    msgToUp1[:,:,1:dispLevels+1] = dataCost + msgFromDown + msgFromRight + msgFromLeft
    msgToDown1[:,:,1:dispLevels+1] = dataCost + msgFromUp + msgFromRight + msgFromLeft
    msgToRight1[:,:,1:dispLevels+1] = dataCost + msgFromUp + msgFromDown + msgFromLeft
    msgToLeft1[:,:,1:dispLevels+1] = dataCost + msgFromUp + msgFromDown + msgFromRight
    
    # Find minimum costs
    minMsgToUp = np.amin(msgToUp1,axis=2)
    minMsgToDown = np.amin(msgToDown1,axis=2)
    minMsgToRight = np.amin(msgToRight1,axis=2)
    minMsgToLeft = np.amin(msgToLeft1,axis=2)

    # Compute messages - Step 2
    for i in range(dispLevels):
        # Messages to up
        costs[:,:,0] = msgToUp1[:,:,i+1]
        costs[:,:,1] = np.minimum(msgToUp1[:,:,i],msgToUp1[:,:,i+2])+lambda_
        costs[:,:,2] = minMsgToUp+2*lambda_
        msgToUp2[:,:,i] = np.amin(costs,axis=2)-minMsgToUp
        
        # Messages to down
        costs[:,:,0] = msgToDown1[:,:,i+1]
        costs[:,:,1] = np.minimum(msgToDown1[:,:,i],msgToDown1[:,:,i+2])+lambda_
        costs[:,:,2] = minMsgToDown+2*lambda_
        msgToDown2[:,:,i] = np.amin(costs,axis=2)-minMsgToDown
        
        # Messages to right
        costs[:,:,0] = msgToRight1[:,:,i+1]
        costs[:,:,1] = np.minimum(msgToRight1[:,:,i],msgToRight1[:,:,i+2])+lambda_
        costs[:,:,2] = minMsgToRight+2*lambda_
        msgToRight2[:,:,i] = np.amin(costs,axis=2)-minMsgToRight
        
        # Messages to left
        costs[:,:,0] = msgToLeft1[:,:,i+1]
        costs[:,:,1] = np.minimum(msgToLeft1[:,:,i],msgToLeft1[:,:,i+2])+lambda_
        costs[:,:,2] = minMsgToLeft+2*lambda_
        msgToLeft2[:,:,i] = np.amin(costs,axis=2)-minMsgToLeft

    # Send messages
    msgFromDown = shiftArray(msgToUp2,[-1,0,0]) #shift up
    msgFromUp = shiftArray(msgToDown2,[1,0,0]) #shift down
    msgFromLeft = shiftArray(msgToRight2,[0,1,0]) #shift right
    msgFromRight = shiftArray(msgToLeft2,[0,-1,0]) #shift left

    # Compute belief
    #belief = dataCost + msgFromUp + msgFromDown + msgFromRight + msgFromLeft #standard belief computation
    belief = msgFromUp + msgFromDown + msgFromRight + msgFromLeft #without dataCost (larger energy but better results)
    
    # Compute the disparity map
    dispMap = np.argmin(belief,axis=2)
    
    # Compute energy
    dataEnergy = np.sum(dataCost[np.arange(rows)[:,np.newaxis],np.arange(cols)[np.newaxis,:],dispMap])
    smoothnessEnergyHorizontal = np.sum(lambda_*np.minimum(np.absolute(np.diff(dispMap,n=1,axis=1)),2))
    smoothnessEnergyVertical = np.sum(lambda_*np.minimum(np.absolute(np.diff(dispMap,n=1,axis=0)),2))
    energy[it] = dataEnergy+smoothnessEnergyHorizontal+smoothnessEnergyVertical

    # Normalize the disparity map for display
    scaleFactor = 256/dispLevels
    dispImg = (dispMap*scaleFactor).astype(np.uint8)

    # Show disparity map
    plt.cla()
    plt.imshow(dispImg,cmap="gray")
    plt.show(block=False)
    plt.pause(0.01)

    # Show energy and iteration
    print("iteration: {0}/{1}, energy: {2}".format(it+1,iterations,energy[it]))

# Show convergence graph
plt.figure()
plt.plot(np.arange(1,iterations+1),energy,marker="o")
plt.xlabel("Iterations")
plt.ylabel("Energy")
plt.show(block=False)
plt.pause(0.01)

# Save disparity map
cv.imwrite("disparityBP_Synch2.png",dispImg)

# Stop timer and display running time
elapsedTime = time.time()-timerVal
print("Running time: {:.2f} seconds".format(elapsedTime))

plt.show()
